buffer/src/main/java/io/netty/buffer/MiMallocByteBufAllocator.java

```diff
@@ -0,0 +1,2409 @@
+        } else if (size < 32 * MiB) {
+            align_size = 1 * MiB;
+        } else {
+            align_size = 4 * MiB;
+        }
+        return alignUp(size, align_size);
+    }
+
+    /**
+     * The strategy for how {@link MiMallocByteBufAllocator} should allocate chunk buffers.
+     */
+    interface ChunkAllocator {
+        /**
+         * Allocate a buffer for a chunk. This can be any kind of {@link AbstractByteBuf} implementation.
+         * @param initialCapacity The initial capacity of the returned {@link AbstractByteBuf}.
+         * @param maxCapacity The maximum capacity of the returned {@link AbstractByteBuf}.
+         * @return The buffer that represents the chunk memory.
+         */
+        AbstractByteBuf allocate(int initialCapacity, int maxCapacity);
+    }
+
+    static class MiByteBuf extends AbstractReferenceCountedByteBuf {
+        private int length;
+        private int maxFastCapacity;
+        private AbstractByteBuf rootParent;
+        private int adjustment;
+        private ByteBuffer tmpNioBuf;
+        private boolean hasArray;
+        private boolean hasMemoryAddress;
+        private Block block;
+
+        MiByteBuf() {
+            super(0);
+        }
+
+        void init(Block block, int length, int maxCapacity) {
+            assert block != null;
+            block.nextBlock = null;
+            this.resetRefCnt();
+            this.discardMarks();
+            this.block = block;
+            this.length = length;
+            this.maxFastCapacity = block.blockBytes;
+            this.adjustment = block.blockAdjustment;
+            maxCapacity(maxCapacity);
+            setIndex0(0, 0);
+            this.rootParent = block.page.segment.delegate;
+            this.tmpNioBuf = null;
+            this.hasArray = block.page.segment.delegate.hasArray();
+            this.hasMemoryAddress = block.page.segment.delegate.hasMemoryAddress();
+        }
+
+        @Override
+        protected void deallocate() {
+            assert this.block != null;
+            this.length = 0;
+            this.maxFastCapacity = 0;
+            this.adjustment = 0;
+            this.rootParent = null;
+            this.tmpNioBuf = null;
+            this.hasArray = false;
+            this.hasMemoryAddress = false;
+            MiMallocByteBufAllocator allocator = this.block.page.segment.parent;
+            Block bk = this.block;
+            this.block = null;
+            allocator.free(bk);
+        }
+
+        public ByteBuf capacity(int newCapacity) {
+            if (length <= newCapacity && newCapacity <= maxFastCapacity) {
+                ensureAccessible();
+                length = newCapacity;
+                return this;
+            }
+            checkNewCapacity(newCapacity);
+            if (newCapacity < capacity()) {
+                length = newCapacity;
+                trimIndicesToCapacity(newCapacity);
+                return this;
+            }
+            // Reallocation required.
```

Comment:
This one need few more comments e.g.

1. the buffer's block is a reusable block node in a free list and represent a defined range of allocation in a page 
2. this buffer can be === the block itself (it's the fast path to avoid creating garbage)
3. due to this we cannot just free block
4. we need to perform a shallow copy of this block (in theory only if `block == this`) and release the copy instead (which will have a shiny new ref cnt, but refer to the original page range)
5. reallocating the internal block makes this `ByteBuf` instance no longer reusable